include/deal.II/numerics/storage.h

```diff
@@ -0,0 +1,243 @@
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_storage_h
+#define dealii_storage_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+
+#include <type_traits>
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A helper class to store a vector of shared pointers to objects of type `T`.
+ * If the number of elements exceeds the specified maximum size of the
+ * container, the oldest element is removed. Additionally, random access and
+ * removal of elements is implemented. Indexing is done relative to the last
+ * added element. In order to optimize the container for usage with
+ * memory-denamding objects (i.e. linear algebra vectors), the removal of an
+ * element does not free the memory. Instead the shared pointer to the element
+ * is being kept in a separate cache so that subsequent addition does not
+ * require re-allocation of memory.
+ *
+ * The primary usage of this class is in solvers to store a history of vectors.
+ * That is, if at the iteration $k$ we store $m$ vectors from previous
+ * iterations
+ * $\{k-1,k-2,...,k-m\}$, then addition of the new element will make
+ * the container contain elements from iterations $\{k,k-1,k-2,...,k-m+1\}$.
+ *
+ * @author Denis Davydov, 2018
+ */
+template <typename T>
+class FiniteSizeHistory
+{
+public:
+  static_assert(std::is_default_constructible<T>::value,
+                "Class T requires default-constructible elements");
+
+  /**
+   * Constructor.
+   *
+   * @parameter max_elements maximum number of elements to be stored in the
+   * queue.
+   */
+  FiniteSizeHistory(const unsigned int max_elements = 0);
+
+  /**
+   * Add new object by copying in front of the queue.
+   * If the maximum number of elements is reached, the oldest element is
+   * removed.
+   */
+  void
+  add(const T &element);
+
+  /**
+   * Remove an element with index @p index,
+   * counting from the last added element.
+   */
+  void
+  remove(const unsigned int index);
+
+  /**
+   * Read/write access to an element with index @p index,
+   * counting from the last added element.
+   */
+  T &operator[](const unsigned int index);
+
+  /**
+   * Read access to an element with index @p index,
+   * counting from the last added element.
+   */
+  const T &operator[](const unsigned int index) const;
+
+  /**
+   * Return the current size of the queue.
+   */
+  unsigned int
+  size() const;
+
+  /**
+   * Return the maximum allowed size of the queue.
+   */
+  unsigned int
+  max_size() const;
+
+  /**
+   * Clear the contents, including the cache.
+   */
+  void
+  clear();
+
+private:
+  /**
+   * Maximum number of elements allowed in the queue.
+   */
+  unsigned int max_n_elements;
+
+  /**
+   * A vector which stores the data.
+   */
+  std::vector<std::shared_ptr<T>> data;
+
+  /**
+   * A vector to cache data, in particular for the case when
+   * removal is followed by addition.
+   */
+  std::vector<std::shared_ptr<T>> cache;
```

Comment:
It might be easier to use `std::deque` here if you push and pop both ends.

Reply:
Oh, I had not realized that `std::dequeue` has random access `operator[]`!